src/core/main.c

```diff
@@ -1253,37 +1253,39 @@ static void bump_file_max_and_nr_open(void) {
 #endif
 
 #if BUMP_PROC_SYS_FS_NR_OPEN
-        int v = INT_MAX;
+        unsigned int v = INT_MAX;
 
-        /* Argh! The kernel enforces maximum and minimum values on the fs.nr_open, but we don't really know
-         * what they are. The expression by which the maximum is determined is dependent on the architecture,
-         * and is something we don't really want to copy to userspace, as it is dependent on implementation
-         * details of the kernel. Since the kernel doesn't expose the maximum value to us, we can only try
-         * and hope. Hence, let's start with INT_MAX, and then keep halving the value until we find one that
-         * works. Ugly? Yes, absolutely, but kernel APIs are kernel APIs, so what do can we do... 🤯 */
+        // cf. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/file.c?h=v6.8#n27
+	#define BITS_PER_LONG __WORDSIZE
+        #define __const_min(x, y) ((x) < (y) ? (x) : (y))
+        unsigned int sysctl_nr_open_max =
+            __const_min(INT_MAX, ~(size_t)0/sizeof(void *)) & -BITS_PER_LONG;
 
         for (;;) {
-                int k;
+                unsigned int k;
+
+                v &= sysctl_nr_open_max;
 
-                v &= ~(__SIZEOF_POINTER__ - 1); /* Round down to next multiple of the pointer size */
                 if (v < 1024) {
                         log_warning("Can't bump fs.nr_open, value too small.");
                         break;
                 }
 
-                k = read_nr_open();
-                if (k < 0) {
-                        log_error_errno(k, "Failed to read fs.nr_open: %m");
+                int nr_open = read_nr_open();
+                if (nr_open < 0) {
+                        log_error_errno(nr_open, "Failed to read fs.nr_open: %m");
                         break;
                 }
+
+                k = (unsigned int)(nr_open);
                 if (k >= v) { /* Already larger */
```

Comment:
given that you only need `k` once, just do the cast to `(unsigned)` right in the comparison?

Reply:
Deal!

Addressed in 1c2bf32eb9bd970e6e592baabfd697f80a418812 .